### PR TITLE
[WIP] API Auth Key

### PR DESCRIPTION
**Pull request by jbrooksuk**
I've been playing around with using an api key for authorisation on the API. I've not fully tested the API yet, but when you supply an incorrect key, you do get an error back.

Needs more work:

- [ ] Generate default api key for users who may not currently have one.
- [x] Generate the api key at setup for the user.
- [x] Check the api still works with basic auth.

### CODE REVIEW

**Comment by jbrooksuk** [comment]
@GrahamCampbell have I registered the service provider properly here?

File: src/Models/User.php [hunk 1]
```diff
@@ -72,4 +73,23 @@ public function getGravatarAttribute($size = 200)
...
         return sprintf('https://www.gravatar.com/avatar/%s?size=%d', md5($this->email), $size);
     }
+
+    /**
+     * Find by api_key, or throw an exception.
+     *
+     * @param string $api_key
+     * @param mixed  $columns
+     *
+     * @throws ModelNotFoundException if no matching User exists.
```
Comment (GrahamCampbell) [inline src/Models/User.php]: should use the fqcn here

File: src/Models/User.php [hunk 2]
```diff
@@ -72,4 +73,23 @@ public function getGravatarAttribute($size = 200)
     {
         return sprintf('https://www.gravatar.com/avatar/%s?size=%d', md5($this->email), $size);
     }
+
+    /**
+     * Find by api_key, or throw an exception.
+     *
+     * @param string $api_key
```
Comment (GrahamCampbell) [inline src/Models/User.php]: should be `$apiKey`

Comment (GrahamCampbell) [inline src/Models/User.php]: laravel should be clever enough to still work

File: src/Providers/AuthServiceProvider.php [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    }
+
+    /**
+     * Register the service provider.
+     *
+     * @return void
+     */
+    public function register()
+    {
+        $this->app->bindShared('CachetHQ\Cachet\Auth\ApiKeyAuthenticator', function ($app) {
```
Comment (GrahamCampbell) [inline src/Providers/AuthServiceProvider.php]: don't need `$app` there. it can be deleted

**Comment by GrahamCampbell** [comment]
Just rebased this and pushed some fixes too.

**Comment by GrahamCampbell** [comment]
NB, you were returning the user id for some reason when the description in the dingo interfaces says you should return an instance of the user.

**Comment by jbrooksuk** [comment]
Thanks buddy.

**Comment by jbrooksuk** [comment]
... Like an idiot I was trying something and committed that too. Will fix when I continue work.

**Comment by GrahamCampbell** [comment]
> have I registered the service provider properly here?

yeh, looks good

**Comment by joecohens** [comment]
:+1:

**Comment by jbrooksuk** [comment]
@GrahamCampbell do you think a `post-update-cmd` as a Laravel command would be good enough, so we can update all existing Heroku deployments who don't have an API key?

**Comment by jbrooksuk** [comment]
I've added a button to the user panel to regenerate the API key anyway, so perhaps we advise that.

**Comment by jbrooksuk** [comment]
Will merge for now, with a note for generating API keys.

**Comment by jbrooksuk** [comment]
I've merged it, but now it's gone from the main code?

**Comment by GrahamCampbell** [comment]
Errrr, what>

**Comment by GrahamCampbell** [comment]
https://github.com/cachethq/Cachet/commits/master

**Comment by GrahamCampbell** [comment]
I can still see it?

**Comment by jbrooksuk** [comment]
I just re-merged it. It wasn't in `master`.

**Comment by GrahamCampbell** [comment]
Grrr. Is this something to do with your git setup. This has happened before?

**Comment by jbrooksuk** [comment]
I think it must be. I'm going to have to check that out.

**Comment by jbrooksuk** [comment]
Are we missing anything else?

**Comment by GrahamCampbell** [comment]
Not sure. Haven't been watching carefully.

**Comment by jbrooksuk** [comment]
I think it's all there.

**Pull request merged**